Added RSS parsing for user feeds

maxminoS opened the pull request:
Uses regular expressions to distinguish feed data, preventing any weird injections in review data if the user decides to for no good reason.

`<b>`, `<em>`, and similar HTML content tags may not be parsed correctly if used by the user.

Added a testing method for the poster and review data.

Use `GetFeed(username)`

maxminoS committed:
Added RSS parsing for user feeds

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	feed, err := fp.ParseURL("https://letterboxd.com/" + username + "/rss/")
+
+	if err != nil {
+		log.Fatalf("failed to fetch feed: %v\n", err)
```
Don't use `Fatalf`, just return an error and let the parent function handle things. We don't want the whole bot to crash just because it failed to get an rss feed.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+
+	entries := []*FeedEntry{}
+	for i := 0; i < 50; i++ {
+		entries = append(entries, parseEntry(feed.Items[i]))
```
What happens if the user has less than 50 entries in their feed?

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+
+	return Feed{
+		username,
+		feed.Items[0].Author.Name,
```
What happens if a user's feed is empty?

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	}
+}
+
+func handleFilmTitle(title []ext.Extension) (string) {
```
If you only have on return type you don't need the `()`, you can just have it like this: `func handleFilmTitle(title []ext.Extension) string {`

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+		return -1
+	}
+
+	i, err := strconv.Atoi(string(rating[0].Value[0]))
+	j, err := strconv.Atoi(string(rating[0].Value[2]))
+
+	if err != nil {
+		return -1
+	}
```
Only the error value from the second cast  is getting checked here, not the first.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	watchedDateString := watchedDate.Format("Monday January 2, 2006")
+
+	// Poster
+	rePoster, _ := regexp.Compile(`^ <p><img src="https://a.ltrbxd.com/resized/(?P<image_url>.+?)"/></p> `)
```
The second return value is an error type, do not ignore it.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	}
+
+	// If spoiler, elseif no review, else review
+	if r, _ := regexp.MatchString(` \(contains spoilers\)$`, title); r {
```
Better to just check if the review starts with `This review may contain spoilers.` instead of using a regex match.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+
+	// If spoiler, elseif no review, else review
+	if r, _ := regexp.MatchString(` \(contains spoilers\)$`, title); r {
+		review = "This review may contain spoilers."
```
Add a new field in `FeedEntry`, `Spoilers bool`. If the review has spoilers we should trim the warning off and then save the review and set `Spoilers` to  `true`. This is just the parser, we shouldn't decide here how we'll present things.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	// If spoiler, elseif no review, else review
+	if r, _ := regexp.MatchString(` \(contains spoilers\)$`, title); r {
+		review = "This review may contain spoilers."
+	} else if r, _ := regexp.MatchString(fmt.Sprintf(` <p>Watched on %s.</p> `, watchedDateString), description); r {
```
Again better to not use regex here. You can use the `WatchedDate` value and format it appropriately to match this string exactly. If there is no `WatchedDate`, then use this as a fallback check. Also, don't ignore errors.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	} else if r, _ := regexp.MatchString(fmt.Sprintf(` <p>Watched on %s.</p> `, watchedDateString), description); r {
+		review = ""
+	} else {
+		reReview, _ := regexp.Compile(` <p>(?P<review>.+)</p> `)
```
Don't ignore errors.

typerslow commented on rss.go:
```diff
@@ -0,0 +1,153 @@
+	watchedDate := handleWatchedDate(entry.Extensions["letterboxd"]["watchedDate"])
+	poster, review := handleData(entry.Title, entry.Description, watchedDate)
+
+	return &FeedEntry{
+		entry.GUID,
+		entry.Link,
+		handleFilmTitle(entry.Extensions["letterboxd"]["filmTitle"]),
+		handleYear(entry.Extensions["letterboxd"]["filmYear"]),
+		handleRating(entry.Extensions["letterboxd"]["memberRating"]),
+		watchedDate,
+		handleRewatch(entry.Extensions["letterboxd"]["rewatch"]),
+		poster,
+		review,
+	}
```
Since there's so many fields here, better to use their names and not just depend on the order. For example:
```go
return &FeedEntry{
	ID: entry.GUID,
	...
}
```

maxminoS committed:
Fixed requested improvements

Cleaner return types;
Specific error handling;
Minimize regexp use;
Cleanly parse HTML;
Added Spoilers to struct;
Created structs for test data;
Cases of empty feed and too few entries;

maxminoS committed:
Fixed other changes

Use cleaner formatting;
Passed policy through program lifecycle;
Fixed test errors;
